core/src/main/java/org/apache/brooklyn/policy/AbstractInvokeEffectorPolicy.java

```diff
@@ -0,0 +1,145 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.brooklyn.policy;
+
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import javax.annotation.Nonnull;
+
+import org.apache.brooklyn.api.effector.Effector;
+import org.apache.brooklyn.api.entity.EntityLocal;
+import org.apache.brooklyn.api.mgmt.Task;
+import org.apache.brooklyn.api.sensor.AttributeSensor;
+import org.apache.brooklyn.config.ConfigKey;
+import org.apache.brooklyn.core.config.ConfigKeys;
+import org.apache.brooklyn.core.policy.AbstractPolicy;
+import org.apache.brooklyn.core.sensor.Sensors;
+import org.apache.brooklyn.util.guava.Maybe;
+import org.apache.brooklyn.util.text.Strings;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Objects;
+import com.google.common.util.concurrent.MoreExecutors;
+
+public abstract class AbstractInvokeEffectorPolicy extends AbstractPolicy {
+
+    private static final Logger LOG = LoggerFactory.getLogger(AbstractInvokeEffectorPolicy.class);
+
+    public static final ConfigKey<String> IS_BUSY_SENSOR_NAME = ConfigKeys.newStringConfigKey(
+            "isBusySensor",
+            "Name of the sensor to publish on the entity that indicates that this policy has incomplete effectors. " +
+                    "If unset running tasks will not be tracked.");
+
+    private final AtomicInteger taskCounter = new AtomicInteger();
+
+    /**
+     * Indicates that onEvent was notified of an value of is not the latest sensor value.
+     */
+    private boolean moreUpdatesComing;
+    /**
+     * The timestamp of the event that informed moreUpdatesComing. Subsequent notifications
+     * of earlier events will not cause updates of moreUpdatesComing.
+     */
+    private long mostRecentUpdate = 0;
+    /**
+     * Guards {@link #moreUpdatesComing} and {@link #mostRecentUpdate}.
+     */
+    private final Object[] moreUpdatesLock = new Object[0];
+
+    @Override
+    public void setEntity(EntityLocal entity) {
+        super.setEntity(entity);
+        if (isBusySensorEnabled()) {
+            // Republishes when the entity rebinds.
+            publishIsBusy();
+        }
+    }
+
+    /**
+     * Invoke effector with parameters on the entity that the policy is attached to.
+     */
+    protected <T> Task<T> invoke(Effector<T> effector, Map<String, ?> parameters) {
+        if (isBusySensorEnabled()) {
+            getTaskCounter().incrementAndGet();
+            publishIsBusy();
+        }
+        Task<T> task = entity.invoke(effector, parameters);
+        if (isBusySensorEnabled()) {
+            task.addListener(new EffectorListener(), MoreExecutors.sameThreadExecutor());
+        }
+        return task;
+    }
+
+    protected boolean isBusy() {
+        synchronized (moreUpdatesLock) {
```

Comment:
I included the synchronisation to make sure that `moreUpdatesComing` is visible to whatever thread invokes the effector's completion listener (which calls `publishIsBusy` which calls `isBusy`). At a minimum the field should be volatile then. Is it a big one for you?